velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,573 @@
+    for (int32_t i = 0; i < numGroups; ++i) {
+      size += value<ComplexTypeAccumulator>(groups[i])->size();
+    }
+    return size;
+  }
+
+  // Combines a partial aggregation represented by the key-value pair at row in
+  // mapKeys and mapValues into groupMap.
+  FOLLY_ALWAYS_INLINE void addToFinalAggregation(
+      const BaseVector* mapKeys,
+      const FlatVector<int64_t>* flatValues,
+      vector_size_t index,
+      const vector_size_t* rawSizes,
+      const vector_size_t* rawOffsets,
+      ComplexTypeAccumulator* accumulator,
+      HashStringAllocator* allocator) {
+    auto size = rawSizes[index];
+    auto offset = rawOffsets[index];
+
+    for (int i = 0; i < size; ++i) {
+      const auto count = flatValues->valueAt(offset + i);
+      accumulator->addValueWithCount(*mapKeys, offset + i, count, allocator);
+    }
+  }
+
+  DecodedVector decodedKeys_;
+  DecodedVector decodedIntermediate_;
+  TypePtr inputType_;
+};
+} // namespace
+
+void registerModeAggregate(
+    const std::string& prefix,
+    bool withCompanionFunctions,
+    bool overwrite) {
+  std::vector<std::shared_ptr<AggregateFunctionSignature>> signatures = {
+      AggregateFunctionSignatureBuilder()
+          .typeVariable("T")
+          .returnType("T")
+          .intermediateType("map(T,bigint)")
+          .argumentType("T")
+          .build(),
+  };
+
+  auto name = prefix + "mode";
+  registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType,
+          const core::QueryConfig& /*config*/) -> std::unique_ptr<Aggregate> {
+        VELOX_CHECK_EQ(
+            argTypes.size(), 1, "{}: unexpected number of arguments", name);
+
+        auto inputType =
+            isRawInput(step) ? argTypes[0] : argTypes[0]->childAt(0);
+        switch (inputType->kind()) {
+          case TypeKind::BOOLEAN:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<bool>>>(resultType);
+          case TypeKind::TINYINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int8_t>>>(resultType);
+          case TypeKind::SMALLINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int16_t>>>(resultType);
+          case TypeKind::INTEGER:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int32_t>>>(resultType);
+          case TypeKind::BIGINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int64_t>>>(resultType);
+          case TypeKind::REAL:
+            return std::make_unique<SimpleAggregateAdapter<ModeAggregate<
+                float,
+                util::floating_point::NaNAwareHash<float>,
+                util::floating_point::NaNAwareEquals<float>>>>(resultType);
+          case TypeKind::DOUBLE:
```

Comment:
Can you add decimal, Date, and unknown type? unknow type is NullType in Spark

Reply:
Date is covered by TypeKind::INT32, Added HUGEINT and UNKNOWN, thanks.